README.rst
======
madmom
======

Madmom is an audio signal processing library written in Python with a strong
focus on music information retrieval (MIR) tasks.

The library is internally used by the Department of Computational Perception,
Johannes Kepler University, Linz, Austria (http://www.cp.jku.at) and the
Austrian Research Institute for Artificial Intelligence (OFAI), Vienna, Austria
(http://www.ofai.at).

Possible acronyms are:

- Madmom Analyzes Digitized Music Of Musicians
- Mostly Audio / Dominantly Music Oriented Modules

It includes reference implementations for some music information retrieval
algorithms, please see the `References`_ section.


Documentation
=============

Documentation of the package can be found online http://madmom.readthedocs.org


License
=======

The package has two licenses, one for source code and one for model/data files.

Source code
-----------

Unless indicated otherwise, all source code files are published under the BSD
license. For details, please see the `LICENSE <LICENSE>`_ file.

Model and data files
--------------------

Unless indicated otherwise, all model and data files are distributed under the
`Creative Commons Attribution-NonCommercial-ShareAlike 4.0
<http://creativecommons.org/licenses/by-nc-sa/4.0/legalcode>`_ license.

If you want to include any of these files (or a variation or modification
thereof) or technology which utilises them in a commercial product, please
contact `Gerhard Widmer <http://www.cp.jku.at/people/widmer/>`_.


Installation
============

Please do not try to install from the .zip files provided by GitHub. Rather
install it from package (if you just want to use it) or source (if you plan to
use it for development) by following the instructions below. Whichever variant
you choose, please make sure that all prerequisites are installed.

Prerequisites
-------------

To install the ``madmom`` package, you must have either Python 2.7 or Python
3.3 or newer and the following packages installed:

- `numpy <http://www.numpy.org>`_
- `scipy <http://www.scipy.org>`_
- `cython <http://www.cython.org>`_
- `nose <https://github.com/nose-devs/nose>`_ (to run the tests)
- `pyaudio <http://people.csail.mit.edu/hubert/pyaudio/>`_ (to process live
  audio input)

If you need support for audio files other than ``.wav`` with a sample rate of
44.1kHz and 16 bit depth, you need ``ffmpeg`` (``avconv`` on Ubuntu Linux has
some decoding bugs, so we advise not to use it!).

Please refer to the `requirements.txt <requirements.txt>`_ file for the minimum
required versions and make sure that these modules are up to date, otherwise it
can result in unexpected errors or false computations!

Install from package
--------------------

The instructions given here should be used if you just want to install the
package, e.g. to run the bundled programs or use some functionality for your
own project. If you intend to change anything within the `madmom` package,
please follow the steps in the next section.

The easiest way to install the package is via ``pip`` from the `PyPI (Python
Package Index) <https://pypi.python.org/pypi>`_::

    pip install madmom

This includes the latest code and trained models and will install all
dependencies automatically.

You might need higher privileges (use su or sudo) to install the package, model
files and scripts globally. Alternatively you can install the package locally
(i.e. only for you) by adding the ``--user`` argument::

    pip install --user madmom

This will also install the executable programs to a common place (e.g.
``/usr/local/bin``), which should be in your ``$PATH`` already. If you
installed the package locally, the programs will be copied to a folder which
might not be included in your ``$PATH`` (e.g. ``~/Library/Python/2.7/bin``
on Mac OS X or ``~/.local/bin`` on Ubuntu Linux, ``pip`` will tell you). Thus
the programs need to be called explicitely or you can add their install path
to your ``$PATH`` environment variable::

    export PATH='path/to/scripts':$PATH

Install from source
-------------------

If you plan to use the package as a developer, clone the Git repository::

    git clone --recursive https://github.com/CPJKU/madmom.git

Since the pre-trained model/data files are not included in this repository but
rather added as a Git submodule, you either have to clone the repo recursively.
This is equivalent to these steps::

    git clone https://github.com/CPJKU/madmom.git
    cd madmom
    git submodule update --init --remote

Then you can simply install the package in development mode::

    python setup.py develop --user

To run the included tests::

    python setup.py test

Upgrade of existing installations
---------------------------------

To upgrade the package, please use the same mechanism (pip vs. source) as you
did for installation. If you want to change from package to source, please
uninstall the package first.

Upgrade a package
~~~~~~~~~~~~~~~~~

Simply upgrade the package via pip::

    pip install --upgrade madmom [--user]

If some of the provided programs or models changed (please refer to the
CHANGELOG) you should first uninstall the package and then reinstall::

    pip uninstall madmom
    pip install madmom [--user]

Upgrade from source
~~~~~~~~~~~~~~~~~~~

Simply pull the latest sources::

    git pull

To update the models contained in the submodule::

    git submodule update

If any of the ``.pyx`` or ``.pxd`` files changed, you have to recompile the
modules with Cython::

    python setup.py build_ext --inplace

Package structure
-----------------

The package has a very simple structure, divided into the following folders:

`/bin <bin>`_
  this folder includes example programs (i.e. executable algorithms)
`/docs <docs>`_
  package documentation
`/madmom <madmom>`_
  the actual Python package
`/madmom/audio <madmom/audio>`_
  low level features (e.g. audio file handling, STFT)
`/madmom/evaluation <madmom/evaluation>`_
  evaluation code
`/madmom/features <madmom/features>`_
  higher level features (e.g. onsets, beats)
`/madmom/ml <madmom/ml>`_
  machine learning stuff (e.g. RNNs, HMMs)
`/madmom/models <../../../madmom_models>`_
  pre-trained model/data files (see the License section)
`/madmom/utils <madmom/utils>`_
  misc stuff (e.g. MIDI and general file handling)
`/tests <tests>`_
  tests

Executable programs
-------------------

The package includes executable programs in the `/bin <bin>`_ folder.
If you installed the package, they were copied to a common place.

All scripts can be run in different modes: in ``single`` file mode to process
a single audio file and write the output to STDOUT or the given output file::

    DBNBeatTracker single [-o OUTFILE] INFILE

If multiple audio files should be processed, the scripts can also be run in
``batch`` mode to write the outputs to files with the given suffix::

    DBNBeatTracker batch [-o OUTPUT_DIR] [-s OUTPUT_SUFFIX] FILES

If no output directory is given, the program writes the output files to the
same location as the audio files.

Some programs can also be run in ``online`` mode, i.e. operate on live audio
signals. This requires `pyaudio <http://people.csail.mit.edu/hubert/pyaudio/>`_
to be installed::

    DBNBeatTracker online [-o OUTFILE] [INFILE]

The ``pickle`` mode can be used to store the used parameters to be able to
exactly reproduce experiments.

Please note that the program itself as well as the modes have help messages::

    DBNBeatTracker -h

    DBNBeatTracker single -h

    DBNBeatTracker batch -h

    DBNBeatTracker online -h

    DBNBeatTracker pickle -h

will give different help messages.


Additional resources
====================

Mailing list
------------

The `mailing list <https://groups.google.com/d/forum/madmom-users>`_ should be
used to get in touch with the developers and other users.

Wiki
----

The wiki can be found here: https://github.com/CPJKU/madmom/wiki

FAQ
---

Frequently asked questions can be found here:
https://github.com/CPJKU/madmom/wiki/FAQ

Citation
========

If you use madmom in your work, please consider citing it:

.. code-block:: latex

   @inproceedings{madmom,
      Title = {{madmom: a new Python Audio and Music Signal Processing Library}},
      Author = {B{\"o}ck, Sebastian and Korzeniowski, Filip and Schl{\"u}ter, Jan and Krebs, Florian and Widmer, Gerhard},
      Booktitle = {Proceedings of the 24th ACM International Conference on
      Multimedia},
      Month = {10},
      Year = {2016},
      Pages = {1174--1178},
      Address = {Amsterdam, The Netherlands},
      Doi = {10.1145/2964284.2973795}
   }

References
==========

.. [1] Florian Eyben, Sebastian Böck, Björn Schuller and Alex Graves,
    *Universal Onset Detection with bidirectional Long Short-Term Memory
    Neural Networks*,
    Proceedings of the 11th International Society for Music Information
    Retrieval Conference (ISMIR), 2010.
.. [2] Sebastian Böck and Markus Schedl,
    *Enhanced Beat Tracking with Context-Aware Neural Networks*,
    Proceedings of the 14th International Conference on Digital Audio Effects
    (DAFx), 2011.
.. [3] Sebastian Böck and Markus Schedl,
    *Polyphonic Piano Note Transcription with Recurrent Neural Networks*,
    Proceedings of the 37th International Conference on Acoustics, Speech and
    Signal Processing (ICASSP), 2012.
.. [4] Sebastian Böck, Andreas Arzt, Florian Krebs and Markus Schedl,
    *Online Real-time Onset Detection with Recurrent Neural Networks*,
    Proceedings of the 15th International Conference on Digital Audio Effects
    (DAFx), 2012.
.. [5] Sebastian Böck, Florian Krebs and Markus Schedl,
    *Evaluating the Online Capabilities of Onset Detection Methods*,
    Proceedings of the 13th International Society for Music Information
    Retrieval Conference (ISMIR), 2012.
.. [6] Sebastian Böck and Gerhard Widmer,
    *Maximum Filter Vibrato Suppression for Onset Detection*,
    Proceedings of the 16th International Conference on Digital Audio Effects
    (DAFx), 2013.
.. [7] Sebastian Böck and Gerhard Widmer,
    *Local Group Delay based Vibrato and Tremolo Suppression for Onset
    Detection*,
    Proceedings of the 13th International Society for Music Information
    Retrieval Conference (ISMIR), 2013.
.. [8] Florian Krebs, Sebastian Böck and Gerhard Widmer,
    *Rhythmic Pattern Modelling for Beat and Downbeat Tracking in Musical
    Audio*,
    Proceedings of the 14th International Society for Music Information
    Retrieval Conference (ISMIR), 2013.
.. [9] Sebastian Böck, Jan Schlüter and Gerhard Widmer,
    *Enhanced Peak Picking for Onset Detection with Recurrent Neural Networks*,
    Proceedings of the 6th International Workshop on Machine Learning and
    Music (MML), 2013.

CHANGES.rst
Release Notes
=============

Version 0.16.dev0
-----------------

New features:

* `TempoDetector` can operate on live audio signals  (#292)
* Added chord evaluation (#309)
* Bar tracking functionality (#316)
* Added `quantize_notes` function (#327)
* Added global key evaluation (#336)
* Added key recognition feature and program (#345, #381)

Bug fixes:

* Fix `TransitionModel` number of states when last state is unreachable (#287)
* Fix double beat detections in `BeatTrackingProcessor` (#298)
* Fix ffmpeg unicode filename handling (#305)
* Fix STFT zero padding (#319)
* Fix memory leak when accessing signal frames (#322)
* Quantization of events does not alter them (#327)

API relevant changes:

* `BufferProcessor` uses `data` instead of `buffer` for data storage (#292)
* `DBNBeatTrackingProcessor` expects 1D inputs (#299)
* Moved downbeat and pattern tracking to `features.downbeats` (#316)
* Write/load functions moved to `io` module (#346)
* Write functions do not return any data (#346)
* Evaluation classes expect annotations/detections, cannot handle files (#346)

Other changes:

* Viterbi decoding of `HMM` raises a warning if no valid path is found (#279)
* Add option to include Nyquist frequency in `STFT` (#280)


Version 0.15.1 (release date: 2017-07-07)
-----------------------------------------

This is a maintenance release.

* NumPy boolean subtract fix (#296)


Version 0.15 (release date: 2017-04-25)
---------------------------------------

New features:

* Streaming mode allows framewise processing of live audio input (#185)
* Exponential linear unit (ELU) activation function (#232)
* `DBNBeatTracker` can operate on live audio signals (#238)
* `OnsetDetectorLL` can operate on live audio signals (#256)

Bug fixes:

* Fix downbeat evaluation failure with a single annotation / detection (#216)
* Fix tempo handling of multi-track MIDI files (#219)
* Fix error loading unicode filenames (#223)
* Fix ffmpeg unicode filename handling (#236)
* Fix smoothing for `peak_picking` (#247)
* Fix combining onsets/notes (#255)

API relevant changes:

* `NeuralNetwork` expect 2D inputs; activation can be computed stepwise (#244)
* Reorder `GRUCell` parameters, to be consistent with all other layers (#243)
* Rename `GRULayer` parameters, to be consistent with all other layers (#243)

Other changes:

* SPL and RMS can be computed on `Signal` and `FramedSignal` (#208)
* `num_threads` is passed to `ParallelProcessor` in single mode (#217)
* Use `install_requires` in `setup.py` to specify dependencies (#226)
* Use new Cython build system to build extensions (#227)
* Allow initialisation of previous/hidden states in RNNs (#243)
* Forward path of `HMM` can be computed stepwise (#244)


Version 0.14.1 (release date: 2016-08-01)
-----------------------------------------

This is a maintenance release.

* `RNNDownBeatProcessor` returns only beat and downbeat activations (#197)
* Update programs to reflect MIREX 2016 submissions (#198)

Version 0.14 (release date: 2016-07-28)
---------------------------------------

New features:

* Downbeat tracking based on Recurrent Neural Network (RNN) and Dynamic
  Bayesian Network (DBN) (#130)
* Convolutional Neural Networks (CNN) and CNN onset detection (#133)
* Linear-Chain Conditional Random Field (CRF) implementation (#144)
* Deep Neural Network (DNN) based chroma vector extraction (#148)
* CRF chord recognition using DNN chroma vectors (#148)
* CNN chord recognition using CRF decoding (#152)
* Initial Windows support (Python 2.7 only, no pip packages yet) (#157)
* Gated Recurrent Unit (GRU) network layer (#167)

Bug fixes:

* Fix downbeat output bug (#128)
* MIDI file creation bug (#166)

API relevant changes:

* Refactored the `ml.rnn` to `ml.nn` and converted the models to pickles (#110)
* Reordered the dimensions of comb_filters to time, freq, tau (#135)
* `write_notes` uses `delimiter` instead of `sep` to separate columns (#155)
* `LSTMLayer` takes `Gate` as arguments, all layers are callable (#161)
* Replaced `online` parameter of `FramedSignalProcessor` by `origin` (#169)

Other changes:

* Added classes for onset/note/beat detection with RNNs to `features.*` (#118)
* Add examples to docstrings of classes (#119)
* Converted `madmom.modules` into a Python package (#125)
* `match_files` can handle inexact matches (#137)
* Updated beat tracking models to MIREX 2015 ones (#146)
* Tempo and time signature can be set for created MIDI files (#166)


Version 0.13.2 (release date: 2016-06-09)
-----------------------------------------

This is a bugfix release.

* Fix custom filterbank in FilteredSpectrogram (#142)

Version 0.13.1 (release date: 2016-03-14)
-----------------------------------------

This is a bugfix release.

* Fix beat evaluation argument parsing (#116)

Version 0.13 (release date: 2016-03-07)
---------------------------------------

New features:

* Python 3 support (3.3+) (#15)
* Online documentation available at http://madmom.readthedocs.org (#60)

Bug fixes:

* Fix nasty unsigned indexing bug (#88)
* MIDI note timing could get corrupted if `note_ticks_to_beats()` was called
  multiple times (#90)

API relevant changes:

* Renamed `DownBeatTracker` and all relevant classes to `PatternTracker` (#25)
* Complete refactoring of the `features.beats_hmm` module (#52)

madmom/io/midi.py
# encoding: utf-8
# pylint: disable=no-member
"""
This module contains MIDI functionality.

"""

from __future__ import absolute_import, division, print_function

import numpy as np
import mido
import warnings

DEFAULT_TEMPO = 500000  # microseconds per quarter note (i.e. 120 bpm in 4/4)
DEFAULT_TICKS_PER_BEAT = 480  # ticks per quarter note
DEFAULT_TIME_SIGNATURE = (4, 4)


# TODO: remove these unit conversion functions after upstream PR is merged
#       https://github.com/olemb/mido/pull/114
def tick2second(tick, ticks_per_beat=DEFAULT_TICKS_PER_BEAT,
                tempo=DEFAULT_TEMPO):
    """
    Convert absolute time in ticks to seconds.

    Returns absolute time in seconds for a chosen MIDI file time resolution
    (ticks/pulses per quarter note, also called PPQN) and tempo (microseconds
    per quarter note).

    """
    # Note: both tempo (microseconds) and ticks are per quarter note
    #       thus the time signature is irrelevant
    scale = tempo * 1e-6 / ticks_per_beat
    return tick * scale


def second2tick(second, ticks_per_beat=DEFAULT_TICKS_PER_BEAT,
                tempo=DEFAULT_TEMPO):
    """
    Convert absolute time in seconds to ticks.

    Returns absolute time in ticks for a chosen MIDI file time resolution
    (ticks/pulses per quarter note, also called PPQN) and tempo (microseconds
    per quarter note).

    """
    # Note: both tempo (microseconds) and ticks are per quarter note
    #       thus the time signature is irrelevant
    scale = tempo * 1e-6 / ticks_per_beat
    return int(round(second / scale))


def bpm2tempo(bpm, time_signature=DEFAULT_TIME_SIGNATURE):
    """
    Convert BPM (beats per minute) to MIDI file tempo (microseconds per
    quarter note).

    Depending on the chosen time signature a bar contains a different number of
    beats. These beats are multiples/fractions of a quarter note, thus the
    returned BPM depend on the time signature.

    """
    return int(round(60 * 1e6 / bpm * time_signature[1] / 4.))


def tempo2bpm(tempo, time_signature=DEFAULT_TIME_SIGNATURE):
    """
    Convert MIDI file tempo (microseconds per quarter note) to BPM (beats per
    minute).

    Depending on the chosen time signature a bar contains a different number of
    beats. These beats are multiples/fractions of a quarter note, thus the
    returned tempo depends on the time signature.

    """
    return 60 * 1e6 / tempo * time_signature[1] / 4.


def tick2beat(tick, ticks_per_beat=DEFAULT_TICKS_PER_BEAT,
              time_signature=DEFAULT_TIME_SIGNATURE):
    """
    Convert ticks to beats.

    Returns beats for a chosen MIDI file time resolution (ticks/pulses per
    quarter note, also called PPQN) and time signature.

    """
    return tick / (4. * ticks_per_beat / time_signature[1])


def beat2tick(beat, ticks_per_beat=DEFAULT_TICKS_PER_BEAT,
              time_signature=DEFAULT_TIME_SIGNATURE):
    """
    Convert beats to ticks.

    Returns ticks for a chosen MIDI file time resolution (ticks/pulses per
    quarter note, also called PPQN) and time signature.

    """
    return int(round(beat * 4. * ticks_per_beat / time_signature[1]))


class MIDIFile(mido.MidiFile):
    """
    MIDI File.

    Parameters
    ----------
    filename : str
        MIDI file name.
    file_format : int, optional
        MIDI file format (0, 1, 2).
    ticks_per_beat : int, optional
        Resolution (i.e. ticks per quarter note) of the MIDI file.
    unit : str, optional
        Unit of all MIDI messages, can be one of the following:

        - 'ticks', 't': use native MIDI ticks as unit,
        - 'seconds', 's': use seconds as unit,
        - 'beats', 'b' : use beats as unit.

    timing : str, optional
        Timing of all MIDI messages, can be one of the following:

        - 'absolute', 'abs', 'a': use absolute timing.
        - 'relative', 'rel', 'r': use relative timing, i.e. delta to
        previous message.

    Examples
    --------
    Create a MIDI file from an array with notes. The format of the note array
    is: 'onset time', 'pitch', 'duration', 'velocity', 'channel'. The last
    column can be omitted, assuming channel 0.

    >>> notes = np.array([[0, 50, 1, 60], [0.5, 62, 0.5, 90]])
    >>> m = MIDIFile.from_notes(notes)
    >>> m  # doctest: +ELLIPSIS
    <madmom.io.midi.MIDIFile object at 0x...>

    The notes can be accessed as a numpy array in various formats (default is
    seconds):

    >>> m.notes
    array([[ 0. , 50. ,  1. , 60. ,  0. ],
           [ 0.5, 62. ,  0.5, 90. ,  0. ]])
    >>> m.unit ='ticks'
    >>> m.notes
    array([[  0.,  50., 960.,  60.,   0.],
           [480.,  62., 480.,  90.,   0.]])
    >>> m.unit = 'seconds'
    >>> m.notes
    array([[ 0. , 50. ,  1. , 60. ,  0. ],
           [ 0.5, 62. ,  0.5, 90. ,  0. ]])
    >>> m.unit = 'beats'
    >>> m.notes
    array([[ 0., 50.,  2., 60.,  0.],
           [ 1., 62.,  1., 90.,  0.]])

    >>> m = MIDIFile.from_notes(notes, tempo=60)
    >>> m.notes
    array([[ 0. , 50. ,  1. , 60. ,  0. ],
           [ 0.5, 62. ,  0.5, 90. ,  0. ]])
    >>> m.unit = 'ticks'
    >>> m.notes
    array([[  0.,  50., 480.,  60.,   0.],
           [240.,  62., 240.,  90.,   0.]])
    >>> m.unit = 'beats'
    >>> m.notes
    array([[ 0. , 50. ,  1. , 60. ,  0. ],
           [ 0.5, 62. ,  0.5, 90. ,  0. ]])

    >>> m = MIDIFile.from_notes(notes, time_signature=(2, 2))
    >>> m.notes
    array([[ 0. , 50. ,  1. , 60. ,  0. ],
           [ 0.5, 62. ,  0.5, 90. ,  0. ]])
    >>> m.unit = 'ticks'
    >>> m.notes
    array([[   0.,   50., 1920.,   60.,    0.],
           [ 960.,   62.,  960.,   90.,    0.]])
    >>> m.unit = 'beats'
    >>> m.notes
    array([[ 0., 50.,  2., 60.,  0.],
           [ 1., 62.,  1., 90.,  0.]])

    >>> m = MIDIFile.from_notes(notes, tempo=60, time_signature=(2, 2))
    >>> m.notes
    array([[ 0. , 50. ,  1. , 60. ,  0. ],
           [ 0.5, 62. ,  0.5, 90. ,  0. ]])
    >>> m.unit = 'ticks'
    >>> m.notes
    array([[  0.,  50., 960.,  60.,   0.],
           [480.,  62., 480.,  90.,   0.]])
    >>> m.unit = 'beats'
    >>> m.notes
    array([[ 0. , 50. ,  1. , 60. ,  0. ],
           [ 0.5, 62. ,  0.5, 90. ,  0. ]])

    >>> m = MIDIFile.from_notes(notes, tempo=240, time_signature=(3, 8))
    >>> m.notes
    array([[ 0. , 50. ,  1. , 60. ,  0. ],
           [ 0.5, 62. ,  0.5, 90. ,  0. ]])
    >>> m.unit = 'ticks'
    >>> m.notes
    array([[  0.,  50., 960.,  60.,   0.],
           [480.,  62., 480.,  90.,   0.]])
    >>> m.unit = 'beats'
    >>> m.notes
    array([[ 0., 50.,  4., 60.,  0.],
           [ 2., 62.,  2., 90.,  0.]])

    """

    UNIT = 'seconds'
    TIMING = 'absolute'

    def __init__(self, filename=None, file_format=0,
                 ticks_per_beat=DEFAULT_TICKS_PER_BEAT, unit=UNIT,
                 timing=TIMING, **kwargs):
        # instantiate a MIDIFile
        super(MIDIFile, self).__init__(filename=filename, type=file_format,
                                       ticks_per_beat=ticks_per_beat, **kwargs)
        # add attributes for unit conversion
        self.unit = unit
        self.timing = timing

    # TODO: remove this method after upstream PR is merged
    #       https://github.com/olemb/mido/pull/115
    def __iter__(self):
        # The tracks of type 2 files are not in sync, so they can
        # not be played back like this.
        if self.type == 2:
            raise TypeError("can't merge tracks in type 2 (asynchronous) file")

        tempo = DEFAULT_TEMPO
        time_signature = DEFAULT_TIME_SIGNATURE
        cum_delta = 0
        for msg in mido.merge_tracks(self.tracks):
            # Convert relative message time to desired unit
            if msg.time > 0:
                if self.unit.lower() in ('t', 'ticks'):
                    delta = msg.time
                elif self.unit.lower() in ('s', 'sec', 'seconds'):
                    delta = tick2second(msg.time, self.ticks_per_beat, tempo)
                elif self.unit.lower() in ('b', 'beats'):
                    delta = tick2beat(msg.time, self.ticks_per_beat,
                                      time_signature)
                else:
                    raise ValueError("`unit` must be either 'ticks', 't', "
                                     "'seconds', 's', 'beats', 'b', not %s." %
                                     self.unit)
            else:
                delta = 0
            # Convert relative time to absolute values if needed
            if self.timing.lower() in ('a', 'abs', 'absolute'):
                cum_delta += delta
            elif self.timing.lower() in ('r', 'rel', 'relative'):
                cum_delta = delta
            else:
                raise ValueError("`timing` must be either 'relative', 'rel', "
                                 "'r', or 'absolute', 'abs', 'a', not %s." %
                                 self.timing)

            yield msg.copy(time=cum_delta)

            if msg.type == 'set_tempo':
                tempo = msg.tempo
            elif msg.type == 'time_signature':
                time_signature = (msg.numerator, msg.denominator)

    def __repr__(self):
        return object.__repr__(self)

    @property
    def tempi(self):
        """
        Tempi (mircoseconds per quarter note) of the MIDI file.

        Returns
        -------
        tempi : numpy array
            Array with tempi (time, tempo).

        Notes
        -----
        The time will be given in the unit set by `unit`.

        """
        # list for all tempi
        tempi = []
        # process all events
        for msg in self:
            if msg.type == 'set_tempo':
                tempi.append((msg.time, msg.tempo))
        # make sure a tempo is set (and occurs at time 0)
        if not tempi or tempi[0][0] > 0:
            tempi.insert(0, (0, DEFAULT_TEMPO))
        # tempo is given in microseconds per quarter note
        # TODO: add otption to return in BPM
        return np.asarray(tempi, np.float)

    @property
    def time_signatures(self):
        """
        Time signatures of the MIDI file.

        Returns
        -------
        time_signatures : numpy array
            Array with time signatures (time, numerator, denominator).

        Notes
        -----
        The time will be given in the unit set by `unit`.

        """
        # list for all time signature
        signatures = []
        # process all events
        for msg in self:
            if msg.type == 'time_signature':
                signatures.append((msg.time, msg.numerator, msg.denominator))
        # make sure a signatures is set (and occurs at time 0)
        if not signatures or signatures[0][0] > 0:
            signatures.insert(0, (0, DEFAULT_TIME_SIGNATURE[0],
                                  DEFAULT_TIME_SIGNATURE[1]))
        # return time signatures
        return np.asarray(signatures, dtype=np.float)

    @property
    def notes(self):
        """
        Notes of the MIDI file.

        Returns
        -------
        notes : numpy array
            Array with notes (onset time, pitch, duration, velocity, channel).

        """
        # lists to collect notes and sustain messages
        notes = []
        sustain_msgs = []
        # dictionary for storing the last onset time and velocity for each
        # individual note (i.e. same pitch and channel)
        sounding_notes = {}

        # as key for the dict use channel * 128 (max number of pitches) + pitch
        def note_hash(channel, pitch):
            """Generate a note hash."""
            return channel * 128 + pitch

        # process all events
        for msg in self:
            # keep track of sustain information
            if msg.type == 'control_change' and msg.control == 64:
                sustain_msgs.append(msg)
            # use only note on or note off events
            note_on = msg.type == 'note_on'
            note_off = msg.type == 'note_off'
            # hash sounding note
            if note_on or note_off:
                note = note_hash(msg.channel, msg.note)
            # start note if it's a 'note on' event with velocity > 0
            if note_on and msg.velocity > 0:
                # save the onset time and velocity
                sounding_notes[note] = (msg.time, msg.velocity)
            # end note if it's a 'note off' event or 'note on' with velocity 0
            elif note_off or (note_on and msg.velocity == 0):
                if note not in sounding_notes:
                    warnings.warn('ignoring MIDI message %s' % msg)
                    continue
                # append the note to the list
                notes.append((sounding_notes[note][0], msg.note,
                              msg.time - sounding_notes[note][0],
                              sounding_notes[note][1], msg.channel))
                # remove hash from dict
                del sounding_notes[note]

        # sort the notes and convert to numpy array
        notes = np.asarray(sorted(notes), dtype=np.float)

        # apply sustain information
        # keep track of sustain start times (channel = key)
        sustain_starts = {}
        note_offsets = notes[:, 0] + notes[:, 2]
        for msg in sustain_msgs:
            # remember sustain start
            if msg.value >= 64:
                sustain_starts[msg.channel] = msg.time
            # expand all notes in this channel until sustain end
            else:
                if msg.channel not in sustain_starts:
                    warnings.warn('ignoring sustain message %s' % msg)
                    continue
                # end all notes with i) offsets between sustain start and end
                sustained = np.logical_and(
                    note_offsets >= sustain_starts[msg.channel],
                    note_offsets <= msg.time)
                # and ii) same channel
                sustained = np.logical_and(sustained,
                                           notes[:, 4] == msg.channel)
                # update duration of notes (sustain end time - onset time)
                notes[sustained, 2] = msg.time - notes[sustained, 0]
                # remove sustain start time for this channel
                del sustain_starts[msg.channel]
        # end all notes latest when next note (of same pitch) starts
        for pitch in np.unique(notes[:, 1]):
            note_idx = np.nonzero(notes[:, 1] == pitch)[0]
            max_duration = np.diff(notes[note_idx, 0])
            notes[note_idx[:-1], 2] = np.minimum(notes[note_idx[:-1], 2],
                                                 max_duration)

        # finally return notes
        return notes

    @classmethod
    def from_notes(cls, notes, unit='seconds', tempo=DEFAULT_TEMPO,
                   time_signature=DEFAULT_TIME_SIGNATURE,
                   ticks_per_beat=DEFAULT_TICKS_PER_BEAT):
        """
        Create a MIDIFile from the given notes.

        Parameters
        ----------
        notes : numpy array
            Array with notes, one per row. The columns are defined as:
            (onset time, pitch, duration, velocity, [channel]).
        unit : str, optional
            Unit of `notes`, can be one of the following:

            - 'seconds', 's': use seconds as unit,
            - 'ticks', 't': use native MIDI ticks as unit,
            - 'beats', 'b' : use beats as unit.

        tempo : float, optional
            Tempo of the MIDI track, given in bpm or microseconds per quarter
            note. The unit is determined automatically by the value:

            - `tempo` <= 1000: bpm
            - `tempo` > 1000: microseconds per quarter note

        time_signature : tuple, optional
            Time signature of the track, e.g. (4, 4) for 4/4.
        ticks_per_beat : int, optional
            Resolution (i.e. ticks per quarter note) of the MIDI file.

        Returns
        -------
        :class:`MIDIFile` instance
            :class:`MIDIFile` instance with all notes collected in one track.

        Notes
        -----
        All note events (including the generated tempo and time signature
        events) are written into a single track (i.e. MIDI file format 0).

        """
        # create new MIDI file
        midi_file = cls(file_format=0, ticks_per_beat=ticks_per_beat,
                        unit=unit, timing='absolute')
        # convert tempo
        if tempo <= 1000:
            # convert from bpm to tempo
            tempo = bpm2tempo(tempo, time_signature)
        else:
            # tempo given in ticks per quarter note
            # i.e. we have to adjust according to the time signature
            tempo = int(tempo * time_signature[1] / 4)
        # create new track and add tempo and time signature information
        track = midi_file.add_track()
        track.append(mido.MetaMessage('set_tempo', tempo=tempo))
        track.append(mido.MetaMessage('time_signature',
                                      numerator=time_signature[0],
                                      denominator=time_signature[1]))
        # create note on/off messages with absolute timing
        messages = []
        for note in notes:
            try:
                onset, pitch, duration, velocity, channel = note
                channel = int(channel)
                velocity = int(velocity)
            except ValueError:
                onset, pitch, duration, velocity = note
                channel = 0
            pitch = int(pitch)
            velocity = int(velocity)
            offset = onset + duration
            # create MIDI messages
            onset = second2tick(onset, ticks_per_beat, tempo)
            note_on = mido.Message('note_on', time=onset, note=pitch,
                                   velocity=velocity, channel=channel)
            offset = second2tick(offset, ticks_per_beat, tempo)
            note_off = mido.Message('note_off', time=offset, note=pitch,
                                    channel=channel)
            # append to list
            messages.extend([note_on, note_off])
        # sort them, convert to relative timing and append to track
        messages.sort(key=lambda msg: msg.time)
        messages = mido.midifiles.tracks._to_reltime(messages)
        track.extend(messages)
        # return MIDI file
        return midi_file

    def save(self, filename):
        """
        Save to MIDI file.

        Parameters
        ----------
        filename : str or open file handle
            The MIDI file name.

        """
        # if we get a filename, open the file
        if not hasattr(filename, 'write'):
            filename = open(filename, 'wb')
        # write the MIDI stream
        self._save(filename)


def load_midi(filename):
    """
    Load notes from a MIDI file.

    Parameters
    ----------
    filename: str
        MIDI file.

    Returns
    -------
    numpy array
        Notes ('onset time' 'note number' 'duration' 'velocity' 'channel')

    """
    return MIDIFile(filename).notes


def write_midi(notes, filename, duration=0.6, velocity=100):
    """
    Write notes to a MIDI file.

    Parameters
    ----------
    notes : numpy array, shape (num_notes, 2)
        Notes, one per row (column definition see notes).
    filename : str
        Output MIDI file.
    duration : float, optional
        Note duration if not defined by `notes`.
    velocity : int, optional
        Note velocity if not defined by `notes`.

    Returns
    -------
    numpy array
        Notes (including note length, velocity and channel).

    Notes
    -----
    The note columns format must be (duration, velocity and channel optional):

    'onset time' 'note number' ['duration' ['velocity' ['channel']]]

    """
    from ..utils import expand_notes
    # expand the array to have a default duration and velocity
    notes = expand_notes(notes, duration, velocity)
    # write the notes to the file and return them
    MIDIFile.from_notes(notes).save(filename)
